src/jit/assertionprop.cpp

```diff
@@ -3384,11 +3394,18 @@ GenTree* Compiler::optAssertionProp_Cast(ASSERT_VALARG_TP assertions, GenTree* t
 {
     assert(tree->gtOper == GT_CAST);
 
-    var_types toType = tree->gtCast.gtCastType;
-    GenTree*  op1    = tree->gtCast.CastOp();
+    var_types fromType = tree->CastFromType();
```

Comment:
I think it would be better to avoid using `CastFromType` as it doesn't do anything useful. You already have the cast operand below so this is simply `var_types fromType = op1->TypeGet()`.

Reply:
Well, I  eventually I want to push the check for IsUnsigned() into CastFromType() 
Currently CastFromType() returns the wrong result when IsUnsigned() is set.

Since this change is going into 3.0 I didn't want to refactor this area more broadly.

Reply:
I'm not sure if it's easy to put the `IsUnsigned` check in `CastFromType`. In some cases (narrowing casts without overflow check) the signedness of the "from" type is not relevant so it should be ignored, otherwise we may end up with different issues. Unfortunately `GT_CAST`'s semantic is extremely complicated (narrowing cast, sign/zero extending cast, FP<->INT cast, overflow checks) and adding a few "helper" functions like `CastFromType` won't necessarily help with that.